netlib/tcp.py

```diff
@@ -484,12 +484,14 @@ def finish(self):
         if not isinstance(self.connection, SSL.Connection):
             if not getattr(self.wfile, "closed", False):
                 try:
-                    self.wfile.flush()
-                    self.wfile.close()
+                    if self.wfile:
```

Comment:
It's triggered if the host mitmproxy is running on, is unable to access the external resource, so it does require iptables to block it:

> Traceback (most recent call last):
  File "/tmp/chal/mitmproxy/mitmproxy/proxy/server.py", line 119, in handle
    root_layer()
  File "/tmp/chal/mitmproxy/mitmproxy/proxy/modes/transparent_proxy.py", line 25, in __call__
    self.disconnect()
  File "/tmp/chal/mitmproxy/mitmproxy/protocol/base.py", line 162, in disconnect
    self.server_conn.finish()
  File "/tmp/chal/mitmproxy/mitmproxy/models/connections.py", line 226, in finish
    tcp.TCPClient.finish(self)
  File "/tmp/chal/mitmproxy/netlib/tcp.py", line 488, in finish
    self.wfile.flush()
AttributeError: 'NoneType' object has no attribute 'flush'

> Traceback (most recent call last):
  File "/tmp/chal/mitmproxy/mitmproxy/proxy/server.py", line 119, in handle
    root_layer()
  File "/tmp/chal/mitmproxy/mitmproxy/proxy/modes/transparent_proxy.py", line 25, in __call__
    self.disconnect()
  File "/tmp/chal/mitmproxy/mitmproxy/protocol/base.py", line 162, in disconnect
    self.server_conn.finish()
  File "/tmp/chal/mitmproxy/mitmproxy/models/connections.py", line 226, in finish
    tcp.TCPClient.finish(self)
  File "/tmp/chal/mitmproxy/netlib/tcp.py", line 494, in finish
    self.rfile.close()
AttributeError: 'NoneType' object has no attribute 'close'